### PR TITLE
Add spreadsheet script to update sheet

### LINKED ISSUE #678: @/Outline VPN Set up
- Create an Outline VPN management app to manage, sending, resending of user VPNs.
- Implement this using Next js and Material UI
- The set up needs to put in mind handing of statistics

### PR DESCRIPTION
**Pull request by koechkevin**
## Description

- Adds a function to update a spreadsheet per email address. 
- The fields we should only be updating are `Outline Key Created?` and  `Key sent`.
- To test, Set up [credentials.json](https://vault.codeforafrica.org/#/vault?search=Credenti&itemId=df036df7-7140-4076-a3f2-d0f40577f79e) and update `.env.local` as per `.env.example`. 
- This PR assumes the sheet provided has the column names as per [this sheet, 1Y9UMaRK0glbvqJ2TUfd4Sjc2zMMCoBKPJeDNrsCvDAw](https://docs.google.com/spreadsheets/d/1Y9UMaRK0glbvqJ2TUfd4Sjc2zMMCoBKPJeDNrsCvDAw/edit#gid=545624287)

Fixes #678 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by thepsalmist — COMMENTED** [review]
LGTM

File: apps/vpnmanager/src/lib/data/spreadsheet.ts [hunk 1]
```diff
@@ -3,10 +3,13 @@ import { google } from "googleapis";
 import { SheetRow } from "@/vpnmanager/types";
 import { toCamelCase } from "@/vpnmanager/utils";
 
+const spreadsheetId = process.env.NEXT_APP_GOOGLE_SHEET_ID;
```
Comment (kilemensi) [inline apps/vpnmanager/src/lib/data/spreadsheet.ts:6]:
```suggestion
const SHEET_ID = process.env.NEXT_APP_GOOGLE_SHEET_ID;
```
There are immutable, correct?

File: apps/vpnmanager/src/lib/data/spreadsheet.ts [hunk 2]
```diff
@@ -45,15 +48,56 @@ async function list(
 }
 
 async function newHires() {
```
Comment (kilemensi) [inline apps/vpnmanager/src/lib/data/spreadsheet.ts:50]: This should probably be `newUsers` or `newAccounts`?

File: apps/vpnmanager/src/lib/data/spreadsheet.ts [hunk 3]
```diff
@@ -45,15 +48,56 @@ async function list(
...
-  const spreadsheetId = process.env.NEXT_APP_GOOGLE_SHEET_ID;
-  const range = process.env.NEXT_APP_GOOGLE_SHEET_RANGE;
   const data = await list(spreadsheetId, range);
   return data.filter(
-    (row: SheetRow) => row.emailAddress && row.keySent !== "Yes",
+    (row: SheetRow) => row.emailAddress && row.keySent?.trim() !== "Yes",
   );
 }
 
+export async function updateSheet(row: Partial<SheetRow>) {
```
Comment (kilemensi) [inline apps/vpnmanager/src/lib/data/spreadsheet.ts:57]: Does this mean we're updating one row at a time? If yes, isn't there a way of reading a range of rows, update them locally and then write the updated range back to gSheet?

Updating one row at a time feels inefficient and unnecessary (unless it's the only way).

**Review by kilemensi — COMMENTED** [review]
👍🏽

**Review by thepsalmist — APPROVED** [review]

File: apps/vpnmanager/src/scripts/processGsheet.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import { processNewUsers } from "@/vpnmanager/lib/processUsers";
```
Comment (kilemensi) [inline apps/vpnmanager/src/scripts/processGsheet.ts:1]: My understanding is scripts aren't really part of the src. They are normally outside the src tree i.e. in the root dir either as `scripts` or `bin`.

File: apps/vpnmanager/src/lib/processUsers.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { SheetRow } from "@/vpnmanager/types";
+import * as Sentry from "@sentry/nextjs";
+
+import spreadsheet, { updateSheet } from "./data/spreadsheet";
```
Comment (kilemensi) [inline apps/vpnmanager/src/lib/processUsers.ts:4]:
```suggestion
import spreadsheet, { updateSheet } from ""@/vpnmanager/lib/data/spreadsheet";
```

File: apps/vpnmanager/src/lib/data/spreadsheet.ts [hunk 6]
```diff
@@ -44,16 +44,67 @@ async function list(
...
 
+function processRow(rows: string[][], row: Partial<SheetRow>) {
+  const { emailAddress, outlineKeyCreated, keySent } = row;
+  const titles = rows[0];
+  const emailIndex = titles.findIndex(
+    (item) => item?.toLowerCase()?.trim() === "email address",
+  );
+  const rowIndexToUpdate = rows.findIndex(
+    (item: string[]) => item[emailIndex] === emailAddress,
+  );
```
Comment (kilemensi) [inline apps/vpnmanager/src/lib/data/spreadsheet.ts:64]: Can rowIndexToUpdate be `-1` i.e. no match for the emailAddress?

File: apps/vpnmanager/src/lib/processUsers.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+export async function processUser(item: SheetRow) {
+  Sentry.captureException(item);
```
Comment (kilemensi) [inline apps/vpnmanager/src/lib/processUsers.ts:7]:
```suggestion
```

You can `log` if you want to see output but this is not an exception.

**Review by kilemensi — APPROVED** [review]

**Pull request merged**